### PR TITLE
Publish Prometheus metrics about build info

### PR DESCRIPTION
**Pull request by pedro-stanaka**
### Summary

I am adding a new metric that I missed while doing my monitoring setup for Keda. This new metric will tell which version of Keda is running in a given Kubernetes cluster. This is really useful when managing multiple clusters to do spot check of which version of Keda is running.

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #

Relates to #4647

### CODE REVIEW

**Review by tomkerkhove — COMMENTED** [review]
LGTM, nice addition - Thanks! Would you mind adding a test for it please, or do we not have tests for metrics yet @JorTurFer?

**Comment by pedro-stanaka** [comment]
> 

I am on the process of adding a test, that is why I left on draft. But running e2e tests on my computer is a PITA. Working on it.

File: pkg/util/build_info.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+package util
+
+import (
+	"github.com/kedacore/keda/v2/pkg/prommetrics"
+	"github.com/kedacore/keda/v2/version"
+	"runtime"
+)
+
+func PublishBuildInfo() {
```
Comment (zroubalik) [inline pkg/util/build_info.go:25]: I am not sure if it is necessary to create a separate package and function for this?

**Review by zroubalik — COMMENTED** [review]
The test can be added here: https://github.com/kedacore/keda/blob/main/tests/sequential/prometheus_metrics/prometheus_metrics_test.go

You can just run a specific test, if you have k8s cluster ready with the KEDA build deployed: https://github.com/kedacore/keda/tree/main/tests#specific-test

File: pkg/util/build_info.go (hunk 1 above)
Comment (pedro-stanaka) [reply pkg/util/build_info.go]: Yeah, good point, inlined the call.

**Comment by pedro-stanaka** [comment]
> The test can be added here: https://github.com/kedacore/keda/blob/main/tests/sequential/prometheus_metrics/prometheus_metrics_test.go
> 
> You can just run a specific test, if you have k8s cluster ready with the KEDA build deployed: https://github.com/kedacore/keda/tree/main/tests#specific-test

Can I change somehow the deploy to use the an image built from my version of the code? Otherwise the test will never pass, since it is running against `ghcr.io/kedacore/keda:main`

File: pkg/prommetrics/prommetrics.go [hunk 2]
```diff
@@ -214,6 +214,6 @@ func DecrementCRDTotal(crdType, namespace string) {
 	crdTotalsGaugeVec.WithLabelValues(crdType, namespace).Dec()
 }
 
-func SetBuildInfo(version, commit, goVersion string) {
-	buildInfo.WithLabelValues(version, commit, goVersion).Set(1)
+func SetBuildInfo(version, commit, goVersion, goos, goarch string) {
```
[the remark this answers is not in the document]
Comment (pedro-stanaka) [reply pkg/prommetrics/prommetrics.go]: Also took this route. thx

**Comment by zroubalik** [comment]
> > The test can be added here: https://github.com/kedacore/keda/blob/main/tests/sequential/prometheus_metrics/prometheus_metrics_test.go
> > You can just run a specific test, if you have k8s cluster ready with the KEDA build deployed: https://github.com/kedacore/keda/tree/main/tests#specific-test
> 
> Can I change somehow the deploy to use the an image built from my version of the code? Otherwise the test will never pass, since it is running against `ghcr.io/kedacore/keda:main`

you can run `make publish && make deploy` just change the image/registry in the Makefile. These commands will build and deploy KEDA on your cluster. Then just run this test.

File: cmd/adapter/main.go [hunk 3]
```diff
@@ -226,6 +227,8 @@ func main() {
 		return
 	}
 
+	prommetrics.SetBuildInfo()
```
Comment (zroubalik) [inline cmd/adapter/main.go:230]: We don't expose metrics on adapter anymore, so you can complete skip this file.

File: pkg/prommetrics/prommetrics.go [hunk 4]
```diff
@@ -204,3 +216,7 @@ func DecrementCRDTotal(crdType, namespace string) {
 
 	crdTotalsGaugeVec.WithLabelValues(crdType, namespace).Dec()
 }
+
+func SetBuildInfo() {
```
Comment (zroubalik) [inline pkg/prommetrics/prommetrics.go:220]:
```suggestion
func RecordBuildInfo() {
```

Also, could you please move this up, right after `RecordScaledObjectError()`, so somewhere around line 188. Please also add a comment with explanation to the function.

File: pkg/prommetrics/prommetrics.go [hunk 5]
```diff
@@ -121,6 +132,7 @@ func init() {
 
 	metrics.Registry.MustRegister(triggerTotalsGaugeVec)
 	metrics.Registry.MustRegister(crdTotalsGaugeVec)
+	metrics.Registry.MustRegister(buildInfo)
```
Comment (zroubalik) [inline pkg/prommetrics/prommetrics.go:135]: Actually, can we just call the new function here and completely skip the call in main.go?

**Review by zroubalik — COMMENTED** [review]
I have some additional thoughts :)

File: cmd/adapter/main.go [hunk 6]
```diff
@@ -175,12 +175,12 @@ func printWelcomeMsg(cmd *Adapter) error {
...
 		return err
 	}
-	version, err := clientset.ServerVersion()
+	k8sVersion, err := clientset.ServerVersion()
 	if err != nil {
 		logger.Error(err, "not able to get Kubernetes version")
 		return err
 	}
-	kedautil.PrintWelcome(logger, kedautil.NewK8sVersion(version), "metrics server")
+	kedautil.PrintWelcome(logger, kedautil.NewK8sVersion(k8sVersion), "metrics server")
```
Comment (zroubalik) [inline cmd/adapter/main.go:183]: this change is no longer needed, could you please skip it?

**Review by zroubalik — COMMENTED** [review]
Great job, could you please fix also this. I will, trigger e2e then! 🙏

File: cmd/adapter/main.go (hunk 6 above)
Comment (pedro-stanaka) [reply cmd/adapter/main.go]: Done

File: pkg/prommetrics/prommetrics.go [hunk 7]
```diff
@@ -173,6 +187,11 @@ func RecordScaledObjectError(namespace string, scaledObject string, err error) {
 	}
 }
 
+// RecordBuildInfo publishes information about Keda version and runtime info through an info metric (gauge).
```
[the remark this answers is not in the document]
Comment (pedro-stanaka) [reply pkg/prommetrics/prommetrics.go]: hahaha, All good. Done as well

**Comment by pedro-stanaka** [comment]
Can one of the maintainers enqueue another e2e run? cc @zroubalik @JorTurFer 

Thanks!

**Review by JorTurFer — APPROVED** [review]

**Comment by JorTurFer** [comment]
/run-e2e sequential/run-e2e sequential/run-e2e sequential/run-e2e sequential/run-e2e sequential/run-e2e sequential/run-e2e sequential

**Comment by JorTurFer** [comment]
Please ignore this:
image

Sometimes I'm like a monkey...
image

**Review by zroubalik — APPROVED** [review]

**Comment by tomkerkhove** [comment]
@JorTurFer Any thoughts why `e2e tests` is still pending?

File: CHANGELOG.md [hunk 8]
```diff
@@ -61,6 +61,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **Pulsar Scaler**: Improve error messages for unsuccessful connections ([#4563](https://github.com/kedacore/keda/issues/4563))
 - **Security:** Enable secret scanning in GitHub repo
 - **RabbitMQ Scaler**: Add support for `unsafeSsl` in trigger metadata ([#4448](https://github.com/kedacore/keda/issues/4448))
+- **Prometheus Metrics**: Add new metric with Keda build info ([#4576](https://github.com/kedacore/keda/pull/4576))
```
Comment (zroubalik) [inline CHANGELOG.md:64]:
```suggestion
- **Prometheus Metrics**: Add new metric with KEDA build info ([#4576](https://github.com/kedacore/keda/pull/4576))
```

We should link an issue here, not the PR. Could you please create one and link it?

**Review by zroubalik — COMMENTED** [review]
We should also document the metric: 
https://keda.sh/docs/2.11/operate/prometheus/

**Comment by pedro-stanaka** [comment]
Yep. Going to.

Pedro Tanaka - Sent from mobile

**Review by zroubalik — APPROVED** [review]
LGTM

Thanks for your patience with all the rounds of review :)

**Comment by pedro-stanaka** [comment]
> We should also document the metric: https://keda.sh/docs/2.11/operate/prometheus/

Docs PR: https://github.com/kedacore/keda-docs/pull/1142

**Pull request merged**